shardingsphere-infra/shardingsphere-infra-binder/src/main/java/org/apache/shardingsphere/infra/binder/segment/insert/values/InsertValueContext.java

```diff
@@ -102,6 +107,10 @@ private int getParameterIndex(final ExpressionSegment valueExpression) {
             }
             if (each instanceof ParameterMarkerExpressionSegment) {
                 result++;
+            } else if (each instanceof BinaryOperationExpression) {
```

Comment:
@CodingBingo Can we extract this common logic?

Reply:
I have thought about this, and I have extract a new method
```
/**
     * calculate parameter count from collections or return index of specific ExpressionSegment.
     *
     * @param assignments expression list
     * @param valueExpression target value expression
     * @return position or index
     */
    public int calculateParameterCountOrIndex(final Collection<ExpressionSegment> assignments, final ExpressionSegment valueExpression) {
        int result = 0;
        for (ExpressionSegment each : assignments) {
            if (valueExpression != null && valueExpression == each) {
                return result;
            }
            if (each instanceof ParameterMarkerExpressionSegment) {
                result++;
            } else if (each instanceof BinaryOperationExpression) {
                if (((BinaryOperationExpression) each).getRight() instanceof ParameterMarkerExpressionSegment) {
                    result++;
                }
            }
        }

        if (valueExpression != null) {
            throw new IllegalArgumentException("Can not get parameter index.");
        }
        return result;
    }
```
But in **InsertValueContext** and **OnDuplicateUpdateContext**,  we both need this method, I am confused about where should I place this common method.

Reply:
@CodingBingo Maybe we can consider adding a tool class to extract all ParameterMarkerExpressionSegment from ExpressionSegment.